clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,142 @@
+* Add an persistent-timer-management attribute to the EJB subsystem timer-service resource (as a alternate to the timer-store attribute) that references a TimerManagementProvider capability from the distributable-ejb subsystem.
+** Modify the TimerService DUP to use a CompositeTimerService in the event that the server is configured with a persistent-timer-management, that delegates to the existing TimerServiceImpl for use with transient timers and the DistributedTimerService for use with persistent timers.
+
+=== Nice-to-Have Requirements
+
+* Add an optional transient-timer-management attribute to the EJB subsystem timer-service resource
+** This would replace the existing TimerService implementation completely, by using a distributed timer service that uses a local, non-persistent cache.
+** This has the advantage of supporting passivation of timer instances to support applications with a sufficiently large number of transient timers in excess of some configured maximum.
+* Add support to TimerServiceMetaData/TimerServiceMetaDataParser for overriding the persistent timer manager per EJB.
+
+=== Non-Requirements
+
+== Backwards Compatibility
+
+* Servers started using an old version of the EJB subsystem configuration will continue to behave as normal, using the existing TimerServiceImpl for both transient and persistent timers.
+
+=== Default Configuration
+
+The following caches will be added to the default ejb cache-container configuration with the Infinispan subsystem:
+
+For the HA profiles:
+
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers:add()
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/component=locking:add(isolation=PESSIMISTIC)
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/component=transaction:add(mode=BATCH)
+	/subsystem=infinispan/cache-container=ejb/distributed-cache=persistent-timers/store=file:add(purge=false)
+
+And for non-HA profiles:
+
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:add()
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:component=locking:add(isolation=PESSIMISTIC)
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers:component=transaction:add(mode=BATCH)
+	/subsystem=infinispan/cache-container=ejb/local-cache=persistent-timers/store=file:add(purge=false,passivation=false)
+
+The following timer management provider will be added to the default distributable-ejb subsystem configuration:
+
+	/subsystem=distributable-ejb/infinispan-timer-management=persistent:add(cache-container=ejb, cache=persistent-timers)
+
+The default timer-service configuration of the EJB subsystem will be modified as follows:
+
+	/subsystem=ejb3/service=timer-service:undefine-attribute(name=default-data-store)
+	/subsystem=ejb3/service=timer-service:write-attribute(name=persistent-timer-management, value=persistent-timers)
```

Comment:
Does that mean timer-service will support both the existing `default-data-store` and the new `persistent-timer-management` attributes, and the infinispan-based `persistent-timer-management` will be the default persistence option?

Reply:
I intended persistent-timer-management to be an alternative attribute to default-data-store, meaning that users can define one or the other, but not both.
When persistent-timer-management is defined, then persistent timers will delegate to a timer manager as defined in the distributable-ejb subsystem.
When a default-data-store is defined instead (meaning that persistent-timer-management must be undefined), the user gets the current behavior.

I don't think we want to change the default configuration within the context of this feature (this can be an independent decision), but if we did, this is what it would look like.

Reply:
All of the timer management services are installed on-demand, so If the timer service is configured to use file or database store in the ejb subsystem, nothing related to the infinispan implementation is started.

Reply:
>  as you stated, changing the default configuration sounds like a separate feature or it should at least be mentioned and documented (e.g.: release notes).

I am not proposing changing the default configuration as part of this feature proposal.

> As far as this analysis document is concerned, I think that it should be clarified that this is a proposal, since it now reads as an implementation detail.

I'm not sure what you mean.  Can you elaborate?